feat: control btc provider state via remote feature flag

fabiobozzo opened the pull request:
## **Description**

Implements Bitcoin provider feature flag support for MetaMask Mobile, mirroring the extension implementation. This allows the New Networks team to control Bitcoin account availability via the `bitcoinAccounts` remote feature flag.

**What is the reason for the change?**
The New Networks team requested Bitcoin provider control via feature flag to match the extension functionality, enabling gradual Bitcoin rollout across both platforms.

**What is the improvement/solution?**
- Wraps Bitcoin provider with `AccountProviderWrapper` for runtime control capability
- Creates `bitcoinAccounts` feature flag selector for UI components
- Updates messenger permissions for future runtime control
- Maintains code fence protection for Bitcoin-specific code
- Provides foundation for runtime feature flag subscription when needed

## **Changelog**

CHANGELOG entry: Added Bitcoin provider feature flag infrastructure for mobile platform

## **Related issues**

[MUL-1139](https://consensyssoftware.atlassian.net/browse/MUL-1139)

## **Manual testing steps**

```gherkin
Feature: Bitcoin provider feature flag support

  Scenario: user creates account with Bitcoin support enabled
    Given Bitcoin feature is built and provider is enabled
    When user onboards with new seed phrase  
    Then account group should include Bitcoin addresses

  Scenario: Bitcoin provider state follows Basic Functionality
    Given Bitcoin accounts are created
    When Basic Functionality is disabled
    Then Bitcoin accounts should be hidden from account lists
```

## **Screenshots/Recordings**

### **Before**
<!-- Bitcoin provider always enabled regardless of feature flag infrastructure -->

### **After**
<!-- Bitcoin provider wrapped with infrastructure ready for runtime flag control -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

[MUL-1139]: https://consensyssoftware.atlassian.net/browse/MUL-1139?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Gates the Bitcoin account provider behind a remote feature flag with runtime subscription, adds a shared enablement helper and UI selector, and updates messengers and tests.
> 
> - **Engine/Service**:
>   - Wrap Bitcoin provider with `AccountProviderWrapper` and initialize `MultichainAccountService` with wrapped provider in `app/core/Engine/controllers/multichain-account-service/multichain-account-service-init.ts`.
>   - Read `RemoteFeatureFlagController:getState`, set initial BTC provider enabled state via `isBitcoinAccountsFeatureEnabled`, subscribe to `RemoteFeatureFlagController:stateChange`, and call `controller.alignWallets()` when enabling.
>   - Expand init messenger permissions to include `RemoteFeatureFlagController:getState` and `RemoteFeatureFlagController:stateChange`; export `Actions`/`Events` in `app/core/Engine/messengers/...`.
> - **Shared feature-flag helper**:
>   - Add `app/multichain-bitcoin/remote-feature-flag.ts` with `isBitcoinAccountsFeatureEnabled` (version-gated via `compare-versions` and `package.json`).
> - **Selectors**:
>   - Add `selectIsBitcoinAccountsEnabled` in `app/selectors/featureFlagController/bitcoinAccountsEnabled/index.ts` using the shared helper.
> - **Tests**:
>   - Update service init tests to use base messenger and assert provider wrapping and runtime enablement behavior.
>   - Add selector tests in `app/selectors/featureFlagController/bitcoinAccountsEnabled/index.test.ts` validating flag and version gating.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d180a4d10680c3835315d88c276eae0712457bb1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
feat: control btc provider state via remote feature flag

Commit:
removed redundant code for Bitcoin feature flag state management simplifying the control of the Bitcoin provider

fabiobozzo committed:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
remove unused RemoteFeatureFlagController actions from multichain account service messenger

Commit:
feat: add Bitcoin provider feature flag infrastructure

- Wrap Bitcoin provider with AccountProviderWrapper for runtime control
- Set initial state based on addBitcoinAccount feature flag
- Integrate with Basic Functionality toggle pattern
- Add feature flag selector for UI components
- Maintain code fence protection for bitcoin builds

aganglada committed:
chore: add bitcoinAccounts feature flag

aganglada committed:
chore: add bitcoinAccounts feature flag

aganglada committed:
chore: move logic out

Commit:
integrate pull/21126 semver comparison

fabiobozzo committed:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
update method for retrieving current version in btc account feature flag

Commit:
fix require package.json syntax

aganglada committed:
chore: fix test

aganglada committed:
Chore/bitcoin feature flag (#21126)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Adds `bitcoinAccounts` feature flag with the following structure

```
{
  "enabled": true,
  "minimumVersion": "7.59.0"
}
```

Adds test for the feature flag and also for the
`MultichainAccountServiceInit`

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce a new version-checked `bitcoinAccounts` feature flag and use
it to enable the BTC provider in Multichain Account Service, replacing
the legacy flag and adding tests.
> 
> - **Feature Flags**:
> - Add `selectIsBitcoinAccountsEnabled` with strict typing and semver
check against app version.
>   - Remove legacy `addBitcoinAccount` selector and its tests.
> - **Engine**:
> - Update `multichain-account-service-init` to read Redux state and
enable `BtcAccountProvider` via `selectIsBitcoinAccountsEnabled`.
> - **Tests**:
> - Add tests for `bitcoinAccountsEnabled` selector version/shape
handling.
> - Expand `multichain-account-service-init` tests to cover BTC provider
enable/disable paths; mock provider wrapper, BTC provider, feature
selector, and Redux state.
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
ddbfc0b25b0c3b8ad9f3b181792344fc4342e372. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

fabiobozzo committed:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

gantunesr commented on app/selectors/featureFlagController/bitcoinAccountsEnabled/index.ts:
```diff
@@ -0,0 +1,53 @@
+
+    if (!assertBitcoinAccountsFeatureFlagType(bitcoinAccountsFlag)) {
+      return false;
+    }
```
@aganglada should the fallback be `true`or `false` in case the flag can't be fetched?

ccharly commented on app/core/Engine/controllers/multichain-account-service/multichain-account-service-init.ts:
```diff
@@ -43,6 +50,16 @@ export const multichainAccountServiceInit: ControllerInitFunction<
+  // - bitcoin feature is built
+  // - bitcoinAccounts feature flag is enabled
+  // - Basic Functionality toggle at runtime (same pattern as Solana)
+  if (selectIsBitcoinAccountsEnabled(ReduxService.store.getState())) {
```
Also, we should not use selectors in engine/controllers contexts, those are for UI contexts.

Instead, we can have a "shared" implementation like:
- https://github.com/MetaMask/metamask-mobile/blob/834e75e592c5af01f7d18ce31b2205dbba632a14/app/multichain-accounts/remote-feature-flag.ts

and use it in both the UI side + in the core/background/app side.

Commit:
address ccharly pr review

Commit:
feat: add Bitcoin account alignment and shared feature flag implementation

Commit:
bugfix cursorbot: mocking mismatch, messenger naming conflict, selector violation

Commit:
lint fixes

Commit:
include alignWallets in testing mock

fabiobozzo committed:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

Commit:
updating the messenger permissions to allow proper access to the RemoteFeatureFlagController

ccharly commented on app/core/Engine/controllers/multichain-account-service/multichain-account-service-init.ts:
```diff
@@ -43,6 +49,28 @@ export const multichainAccountServiceInit: ControllerInitFunction<
+      );
+    });
+  }
+  /// END:ONLY_INCLUDE_IF
```
We miss the "reactive part" when `RemoteFeatureFlagController:stateChange` gets called no?

fabiobozzo committed:
Update app/selectors/featureFlagController/bitcoinAccountsEnabled/index.ts

Co-authored-by: Charly Chevalier <charlyy.chevalier@gmail.com>

Commit:
restore stateChange subscriber

ccharly commented on app/core/Engine/controllers/multichain-account-service/multichain-account-service-init.test.ts:
```diff
@@ -55,6 +104,47 @@ describe('MultichainAccountServiceInit', () => {
+      expect(mockSetEnabled).not.toHaveBeenCalled();
+    });
+
+    it('enables Bitcoin provider when feature flag is enabled and version meets minimum', () => {
```
```suggestion
    it('enables Bitcoin provider when feature flag is enabled', () => {
```

ccharly commented on app/core/Engine/controllers/multichain-account-service/multichain-account-service-init.test.ts:
```diff
@@ -55,6 +90,46 @@ describe('MultichainAccountServiceInit', () => {
+      // Then Bitcoin provider should be enabled
+      expect(mockSetEnabled).toHaveBeenCalledTimes(1);
+      expect(mockSetEnabled).toHaveBeenCalledWith(true);
+    });
```
We could indeed mock `RemoteFeatureFlagController:getState` directly, so we really rely on the `isBitcoinAccountsFeatureEnabled` call logic.

I'm ok with both options, the `:getState` is slightly more correct though.

WDYT @fabiobozzo?

ccharly commented on app/core/Engine/controllers/multichain-account-service/multichain-account-service-init.ts:
```diff
@@ -43,6 +49,62 @@ export const multichainAccountServiceInit: ControllerInitFunction<
+        'Failed to align wallets after enabling Bitcoin provider:',
+        error,
+      );
+    });
```
Actually, just thought of something (which is also true for extension, but potentially will less impact).

This logic here, means that every time we start the app, we'll run an alignment, and the more accounts we have, the longer this will be running. Also, alignment is a mutable operation, thus, this will block any other wallet operations until this one succeed...

Maybe we should improve the alignment logic and add a `isAligned` method on each providers that would check (as fast as possible) if a group is "aligned" for a given provider.

e.g:
- Bitcoin would check if a multichain account group contains N Bitcoin accounts

This operation would be "sync" and could be run before "locking" the wallet mutex!

ccharly committed:
chore: revert app/store/migrations/index.ts

ccharly committed:
refactor: remove dist import

ccharly commented on app/core/Engine/messengers/multichain-account-service-messenger/multichain-account-service-messenger.ts:
```diff
@@ -35,8 +35,6 @@ type Actions =
   | KeyringControllerStateChangeEvent
   | AccountsControllerAccountAddedEvent
-  | AccountsControllerAccountRemovedEvent
-  | AccountsControllerAccountAddedEvent
```
Actually they were duplicated for some reasons 🙃 that's ok to remove them now!

ccharly commented on app/core/Engine/messengers/multichain-account-service-messenger/multichain-account-service-messenger.ts:
```diff
@@ -35,8 +37,6 @@ type Actions =
   | KeyringControllerStateChangeEvent
   | AccountsControllerAccountAddedEvent
-  | AccountsControllerAccountRemovedEvent
-  | AccountsControllerAccountAddedEvent
```
For future reviewers: Those were duplicates.

ccharly committed:
refactor: use proper state type

ccharly committed:
test: fix test

ccharly committed:
fix: fix lint:tsc

ccharly committed:
feat: control btc provider state via remote feature flag (#21345)

## **Description**

Some fixes for:
- https://github.com/MetaMask/metamask-mobile/pull/20989

## **Changelog**

CHANGELOG entry: null

## **Related issues**

N/A

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

### **Before**

### **After**

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Wire BTC provider enable/disable and wallet alignment to the remote
feature flag controller, add typed init messaging, and update
selectors/tests accordingly.
> 
> - **Engine**:
> - **MultichainAccountService init**: Reads
`RemoteFeatureFlagController:getState` and subscribes to
`RemoteFeatureFlagController:stateChange` (typed) to enable/disable
`btcProvider` and trigger `alignWallets` when
`remoteFeatureFlags.bitcoinAccounts` is enabled.
> - **Messaging**:
> - Update init messenger to allow
`RemoteFeatureFlagController:getState` and
`RemoteFeatureFlagController:stateChange` using exported types, removing
ad-hoc/unknown typing.
> - **Selectors**:
> - Improve typing for `selectIsBitcoinAccountsEnabled` by handling
`remoteFeatureFlags` explicitly; strengthen test types and cases.
> - **Tests**:
> - Refactor `multichain-account-service-init` tests to mock
`RemoteFeatureFlagController:getState`, use spies on
`AccountProviderWrapper.setEnabled` and
`MultichainAccountService.alignWallets`, and cover
enabled/disabled/min-version scenarios.
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
c4341f11d5fbe36e3c392d6271003e58fdb1a1ae. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

gantunesr committed:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

ccharly committed:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

ccharly committed:
test: add test for :stateChange

ccharly committed:
test: move event subscriptions/regiters

ccharly committed:
refactor: remove unused null-checks

ccharly committed:
Merge branch 'main' into feat/MUL-1138-ff-disable-btc-provider

ccharly committed:
fix: fix initial logic
